modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/ConnectionCheckTest.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.discovery.tcp;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.Socket;
+import java.util.concurrent.Exchanger;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.util.GridTestClockTimer;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryAbstractMessage;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryConnectionCheckMessage;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryNodeFailedMessage;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static com.sun.tools.javac.util.Assert.check;
+
+/**
+ * Checks pinging next node in the ring relies on configured timeouts.
+ */
+public class ConnectionCheckTest extends GridCommonAbstractTest {
+    /** Number of the ping messages to ensure node pinging works well. */
+    private static final int PING_MESSAGES_CNT_TO_ENSURE = 15;
+
+    /** Timer granulation in milliseconds. See {@link GridTestClockTimer}. */
+    private static final int TIMER_GRANULATION = 10;
+
+    /**
+     * Maximal additional delay before sending the ping message including timer granulation in and other delays
+     * like code delays and/or GC.
+     */
+    private static final int ACCEPTABLE_ADDITIONAL_DELAY = TIMER_GRANULATION + 50;
```

Comment:
magic number

Reply:
> magic number

There is a comment: "Maximal additional delay before sending the ping message including timer granulation in and other delays like code delays and/or GC"

We always can be hit by GC pause, timer granulation, delays in other code. Not giving additional time to wait can lead to flaky tests. Makes sense?